Implement anxcloud_vlan datasource

nachtjasmin opened the pull request:
### Description

Closes SYSENG-1281.

The schema is quite hacky and should definitely be reviewed. It is working, it's just not "clean". ^^

The VLAN data source test can be run with the following command:

	# requires that the ANEXIA_TOKEN is set in the environment
	TF_ACC=true go test ./anxcloud/ -run TestAccAnxCloudVLANDataSource

<!--- Please leave a helpful description of the pull request here. --->

### Checklist

* [x] added release notes to `Unreleased` section in [CHANGELOG.md](CHANGELOG.md), if user facing change
* [ ] tackle codeclimate (*sighs*)
* [x] merging schemas if possible

### References

<!---
Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Vendor blog posts or documentation?
--->
### Community Note
<!--- Please keep this note for the community --->
* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

anx-mschaefer commented on anxcloud/data_source_vlan.go:
```diff
@@ -0,0 +1,111 @@
+		Description: "Provides details about an Anexia Cloud VLAN. This data source is useful if you want to use a non-terraform managed VLAN.",
+		ReadContext: dataSourceVLANRead,
+
+		// TODO: can I reuse the existing schema?
```
we do have some helper functions to override field behaviors (see https://github.com/anexia-it/terraform-provider-anxcloud/blob/main/anxcloud/utils.go#L18)
but if the schemas are "too different", I tend to define them separately

anx-mschaefer commented on anxcloud/data_source_vlan.go:
```diff
@@ -0,0 +1,111 @@
+			return v.List(ctx, page, 100, url.QueryEscape(name))
+		})
+		if err != nil {
+			return diag.FromErr(err)
```
Please wrap `err` with a suitable message. Engine messages can sometimes be quite vague.

anx-mschaefer commented on anxcloud/data_source_vlan.go:
```diff
@@ -0,0 +1,111 @@
+
+	// Once we determined our identifier, we set it and reuse our code for fetching the detailed information.
+	d.SetId(foundVLANID)
+	if err := resourceVLANRead(ctx, d, m); err != nil {
+		return err
+	}
```
```suggestion
	if diags := resourceVLANRead(ctx, d, m); diags.HasError() {
		return diags
	}
```

anx-mschaefer commented on anxcloud/data_source_vlan_test.go:
```diff
@@ -0,0 +1,46 @@
+)
+
+func TestAccAnxCloudVLANDataSource(t *testing.T) {
+	// Note: The VLAN for testing can be accessed here: https://engine.anexia-it.com/ipam/prefix/relation/00a239d617504e4ab49122efe0d27657
```
I think instead of hard-coding the identifier and name, we should use the VLAN that is already passed in via env vars so that we can change them without updating the code, in case someone deletes them by accident^^

Retrieve the VLANs ID:
```go
	environment.SkipIfNoEnvironment(t)
	vlanID := environment.GetEnvInfo(t).VlanID
```

To resolve VLAN name:
```go
	c := testAccProvider.Meta().(providerContext).legacyClient
	v := vlan.NewAPI(c)
	v.Get ...
```

anx-mschaefer reviewed (changes requested):
Already very good! :+1:
Added some comments.

nachtjasmin commented:
I refactored the code a bit and now it's ID > name. I think that makes the actual code a bit easier to understand, although Codeclimate is still not happy about the high number of returns. :/

Therefore I added just another commit which splits up the code a bit. Not sure whether that actually helps with readability, but if it satisfies the linter, well, yeah, let's do it then.

anx-mschaefer commented on anxcloud/data_source_vlan.go:
```diff
@@ -0,0 +1,118 @@
+		}
+	}
+
+	// If the ID is still empty, that means that we haven't found a suitable VLAN.
+	if foundID == "" {
+		return "", diag.Errorf("No VLAN found with the name %q.", name)
+	}
+
+	return "", nil
```
This could reduce the number of returns
```suggestion
	var diags diag.Diagnostics

	// If the ID is still empty, that means that we haven't found a suitable VLAN.
	if foundID == "" {
		diags = append(diags, diag.Errorf("No VLAN found with the name %q.", name)...)
	}

	return foundID, diags
```

anx-mschaefer commented:
I think your changes in `provider.go` got lost while refactoring. That's why code lint fails.

Please also add an example and generate docs for that data-source.

nachtjasmin commented:
> I think your changes in provider.go got lost while refactoring. That's why code lint fails.

oh heck, sorry for that! And thanks for the suggestion with the diagnostics, now CodeClimate should be happy. :+1:

anx-mschaefer commented:
LGTM :+1:
